spot_driver_cpp/src/interfaces/spot_interface.cpp

```diff
@@ -0,0 +1,335 @@
+// Copyright (c) 2023 Boston Dynamics AI Institute LLC. All rights reserved.
+
+#include <spot_driver_cpp/interfaces/spot_interface.hpp>
+
+#include <bosdyn/api/directory.pb.h>
+#include <bosdyn/api/image.pb.h>
+#include <bosdyn/client/gripper_camera_param/gripper_camera_param_client.h>
+#include <builtin_interfaces/msg/time.hpp>
+#include <cv_bridge/cv_bridge.h>
+#include <google/protobuf/duration.pb.h>
+#include <google/protobuf/timestamp.pb.h>
+#include <opencv2/imgcodecs.hpp>
+#include <sensor_msgs/distortion_models.hpp>
+#include <sensor_msgs/msg/camera_info.hpp>
+#include <sensor_msgs/msg/image.hpp>
+#include <spot_driver_cpp/spot_image_sources.hpp>
+#include <spot_driver_cpp/types.hpp>
+#include <std_msgs/msg/header.hpp>
+#include <tl_expected/expected.hpp>
+
+#include <algorithm>
+#include <iostream>
+#include <stdexcept>
+#include <utility>
+
+namespace
+{
+tl::expected<int, std::string> getCvPixelFormat(const bosdyn::api::Image_PixelFormat& format)
+{
+  switch(format)
+  {
+    case bosdyn::api::Image_PixelFormat_PIXEL_FORMAT_RGB_U8:
+    {
+      return CV_8UC3;
+    }
+    case bosdyn::api::Image_PixelFormat_PIXEL_FORMAT_RGBA_U8:
+    {
+      return CV_8UC4;
+    }
+    case bosdyn::api::Image_PixelFormat_PIXEL_FORMAT_GREYSCALE_U8:
+    {
+      return CV_8UC1;
+    }
+    case bosdyn::api::Image_PixelFormat_PIXEL_FORMAT_GREYSCALE_U16:
+    {
+      return CV_16UC1;
+    }
+    case bosdyn::api::Image_PixelFormat_PIXEL_FORMAT_DEPTH_U16:
+    {
+      return CV_16UC1;
+    }
+    default:
+    {
+      return tl::make_unexpected("Unknown pixel format.");
+    }
+  }
+}
+
+builtin_interfaces::msg::Time applyClockSkew(const google::protobuf::Timestamp& timestamp, const google::protobuf::Duration& clock_skew)
+{
+  long int seconds_unskewed = timestamp.seconds() - clock_skew.seconds();
+  int nanos_unskewed = timestamp.nanos() - clock_skew.nanos();
+
+  // Carry over a second if needed
+  // Note: Since ROS Time messages store the nanoseconds component as an unsigned integer, we need to do this before converting to ROS Time.
+  if (nanos_unskewed < 0)
+  {
+    nanos_unskewed += 1e9;
+    seconds_unskewed -= 1;
+  }
+  else if (nanos_unskewed >= 1e9)
+  {
+    nanos_unskewed -= 1e9;
+    seconds_unskewed += 1;
+  }
+
+  // If the timestamp contains a negative time, create an all-zero ROS Time.
+  if (seconds_unskewed < 0)
+  {
+    return builtin_interfaces::build<builtin_interfaces::msg::Time>().sec(0).nanosec(0);
```

Comment:
Can I get a definition on the clock_skew? And does this specific case happen because the spot api time is running before the ROS driver?

Reply:
The Boston Dynamics Spot SDK docs have a section about the clock sync service: https://dev.bostondynamics.com/docs/concepts/base_services#time-sync

I'll add a doc comment describing what this does more comprehensively.

This specific line is to handle an edge case if the Spot API time is running ahead of the driver PC's system clock AND the epoch is very small. In practice I wouldn't expect this to happen, but it's a possibility due to the differences between ROS and Spot SDK time types.

Reply:
Yeah a little comment/documentation just say which time reference we are going from -> to would be helpful here. Also if the Spot time is always running ahead would this mean the ROS Time would constantly be behind or do we reset the spot time somewhere to sink them?

Reply:
> It should only get set to zero if the timestamp is smaller than the clock skew and the timestamp is epoch time right? So that should basically never happen? Just checking my understanding.

This is my understanding as well -- I would only expect this to happen if both Spot and the host PC think it's very early in the morning on January 1, 1970.